### PR TITLE
dev(ci): complete refactor

### PR DESCRIPTION
**Pull request by oxy**
Slowly reorganize CI steps and switch to using GH actions instead of docker images where possible. Initial test only replaces the `fmt` runner.

### CODE REVIEW

File: ci/steps/nfpm.sh
Comment (oxy) [inline ci/steps/nfpm.sh:5]: https://github.com/goreleaser/nfpm/issues/314

File: .github/workflows/ci.yaml
Comment (jsjoeio) [inline .github/workflows/ci.yaml:12]: 👏

File: .github/workflows/ci.yaml
Comment (jsjoeio) [inline .github/workflows/ci.yaml:15]: 👏

File: .github/workflows/ci.yaml
Comment (jsjoeio) [inline .github/workflows/ci.yaml:25]:
```suggestion
      - name: Install Node.js v12
```

File: .github/workflows/ci.yaml
Comment (jsjoeio) [inline .github/workflows/ci.yaml:85]: Genius! 🔥

File: .github/workflows/ci.yaml
Comment (jsjoeio) [inline .github/workflows/ci.yaml:87]: What's this for?

File: .github/workflows/ci.yaml
Comment (jsjoeio) [inline .github/workflows/ci.yaml:144]:
```suggestion
      - name: Install Node.js v12
```

File: .github/workflows/ci.yaml
Comment (jsjoeio) [inline .github/workflows/ci.yaml:149]: What's this used for? I had a note to myself to see if we could remove it

File: .github/workflows/ci.yaml
Comment (jsjoeio) [inline .github/workflows/ci.yaml:166]: I wonder if this should be grouped with the testing workflow

File: .github/workflows/ci.yaml
Comment (jsjoeio) [inline .github/workflows/ci.yaml:191]:
```suggestion
      - name: Install Node.js v12
```

File: .github/workflows/ci.yaml
Comment (jsjoeio) [inline .github/workflows/ci.yaml:215]: How come we do this? (might be helpful to add a comment above)

File: .github/workflows/ci.yaml
Comment (jsjoeio) [inline .github/workflows/ci.yaml:270]: OOO interesting.. bug in playwright?

File: package.json
Comment (jsjoeio) [inline package.json:68]: How come we're adding this in?

**Review by jsjoeio — CHANGES_REQUESTED** [review]
Great work! Initial pass looks 🔥

A couple things:
- thank you for adding comments! 
- nit: can we use spell Node with a capital "N"? 😛
- can we split up workflows for main, release, testing? https://github.com/cdr/code-server/issues/2903 that way, if a test fails, we only have to re-run those jobs

**Comment by oxy** [comment]
> * thank you for adding comments!

:smile: 

> * nit: can we use spell Node with a capital "N"? :stuck_out_tongue:

Sure! patching that soon.

> * can we split up workflows for main, release, testing? #2903 that way, if a test fails, we only have to re-run those jobs

I'm hoping that since the CI is much faster in general (lint/fmt now only takes 3min), and since we cache wherever possible right now (currently yarn build dependencies, VSCode build - will refactor all the scripts in `ci/` in a future PR once this lands, and also cache the `package` steps), we shouldn't strictly need to split it into multiple workflows. In addition, I'm not sure if its possible for items in one workflow to depend on another.

File: .github/workflows/ci.yaml
Comment (oxy) [reply .github/workflows/ci.yaml:270]: "bug"-ish - it sometimes writes browser binaries to `~/.cache/`, sometimes doesn't, and sometimes it looks for it in there and can't find it.

Wasn't sure about adding home-relative directories to our cache, so I just threw that in there.

File: package.json
Comment (oxy) [reply package.json:68]: `shellcheck` is also offered as an npm package that downloads the latest release from GitHub - see https://github.com/gunar/shellcheck/blob/master/download.js - and I figured that this would make it easier for writing CI scripts, and also for contributors, since `shellcheck` would just automatically be pulled and live in `node_modules` instead of having to install it to the entire system.

File: .github/workflows/ci.yaml
Comment (oxy) [reply .github/workflows/ci.yaml:215]: It's all the magic above - I set `AR`/`CC`/`CXX`/`LINK` to their `aarch64` cross compile variants in the job's `env`, and then yarn would use those to build; however qemu-user-static doesn't correctly run node aarch64 on 16.04 (tested locally), so I just sub in the "native"/x86-64 node with the arm64 node before we package.

File: .github/workflows/ci.yaml
Comment (oxy) [reply .github/workflows/ci.yaml:166]: I'm not sure if we should move this to the e2e workflow - this test essentially downloads and installs a vscode theme extension, just as a very quick test to see if `code-server` can run correctly. What do you think we should do?

File: .github/workflows/ci.yaml
Comment (oxy) [reply .github/workflows/ci.yaml:149]: `nfpm` is what we're using right now to build `.deb`/`.rpm` files - it may be possible to move off nfpm, but I think that's better suited for another PR down the line :p

File: .github/workflows/ci.yaml
Comment (oxy) [reply .github/workflows/ci.yaml:87]: this - the hash of the last commit that touched a file inside `lib/vscode` - is the "key" that I'm using to cache the vscode build - it takes about 5-10 minutes in CI, so if we've built it once before and no files inside changed, just pull the build from the cache and then package it up for the next step.

to check history, I needed to set `fetch-depth: 0` in `actions/checkout`, but a full clone instead of a shallow clone only costs us about three to five seconds right now.

Comment (jsjoeio) [reply .github/workflows/ci.yaml:87]: Ahh, I see now. That makes sense! Might be worth adding what you wrote as part of your comment so Future us won't forget!

File: .github/workflows/ci.yaml
Comment (jsjoeio) [reply .github/workflows/ci.yaml:149]: Sounds good to me!

File: .github/workflows/ci.yaml
Comment (jsjoeio) [reply .github/workflows/ci.yaml:166]: I'll lean towards your judgement! If you think it's fine here, let's leave it. We can always move it later if needed!

File: .github/workflows/ci.yaml
Comment (jsjoeio) [reply .github/workflows/ci.yaml:215]: Alrighty then! Might be worth adding a comment here just summarizing that.

File: .github/workflows/ci.yaml
Comment (jsjoeio) [reply .github/workflows/ci.yaml:270]: Argh that's a bummer! Well we can always look into it later. Glad you found a workaround!

File: package.json
Comment (jsjoeio) [reply package.json:68]: Ah gotcha. Sounds good to me!

**Comment by jsjoeio** [comment]
> I'm hoping that since the CI is much faster in general (lint/fmt now only takes 3min), and since we cache wherever possible right now (currently yarn build dependencies, VSCode build - will refactor all the scripts in ci/ in a future PR once this lands, and also cache the package steps), we shouldn't strictly need to split it into multiple workflows. In addition, I'm not sure if its possible for items in one workflow to depend on another.

Ah, I see what you mean. Yeah not sure if that's possible either. Okay no worries then!

**Review by code-asher — DISMISSED** [review]
I haven't been able to do a full review yet, just dropping this here so we remember to get the libc dependencies back down.

File: .github/workflows/ci.yaml
Comment (jsjoeio) [reply .github/workflows/ci.yaml:226]: Thoughts on adding a comment here?

**Review by jsjoeio — COMMENTED** [review]
Nice work 🔥

Left one comment asking you to document a couple lines in the CI file. I'll let @code-asher give the final approval

File: .github/workflows/ci.yaml
Comment (oxy) [reply .github/workflows/ci.yaml:226]: I thought that it made the most sense to explain the situation above the whole workflow, before I set the env vars etc., so I put it here - https://github.com/cdr/code-server/pull/2966/files#diff-944291df2c9c06359d37cc8833d182d705c9e8c3108e7cfe132d61a06e9133ddR181-R187 (you might need to click `Load diff` for the lines to get highlighted)

Comment (jsjoeio) [reply .github/workflows/ci.yaml:226]: Ah, I missed that! LGTM!

**Review by jsjoeio — APPROVED** [review]
Wait for second approval but LGTM!

**Comment by oxy** [comment]
Hi there! This has been stalling for a while, mostly because building against an old version of glibc for CentOS 7 on a newer version of Ubuntu turns out to be not an easy process.

I haven't pushed these experiments, but I know they don't work:
- [glibc_version_header](https://github.com/wheybags/glibc_version_header) - repo says it doesn't work with libstdc++ easily
- `-static-libgcc -static-libstdc++ -static` - does generate `.o` files instead of `.so` files but then `node-gyp` creates a shared object after the fact anyway :/
- `statifer` on the `.node` led to weird bugs on my setup
- I also looked at the `node-gyp` source code to see if I can force static `.node` files, but I felt lost in there so I bailed after a few hours.

Over the next days, I'll be looking at docker buildx, and at "dropping" compatibility for CentOS 7 for per-pull request CI and moving the slower CentOS 7 build to a tag/draft release workflow.

I really hope we can get new CI soon - I've been fighting this for weeks x.x

**Comment by jsjoeio** [comment]
No worries at all! Thanks for the update.

> Over the next days, I'll be looking at docker buildx, and at "dropping" compatibility for CentOS 7 for per-pull request CI and moving the slower CentOS 7 build to a tag/draft release workflow.

Is it possible to break up this work? i.e. get this PR to a stable state, review, merge and then do the remaining work in a follow-up PR? Just wondering if that would help at all.

**Comment by oxy** [comment]
We definitely could do that - the one thing I'm worried about is that we land this (and temporarily drop CentOS 7 support), and then we end up stalling on a follow-up that refactors our release workflow and gets CentOS 7 working again, and having to delay a release/manually build+upload binaries - given our v3.9.2 incident, I don't really want to repeat that >.<

**Comment by jsjoeio** [comment]
> given our v3.9.2 incident, I don't really want to repeat that >.<

Me neither! Sounds good to me then holding off.

**Comment by jsjoeio** [comment]
@oxy  did you address this?

> just dropping this here so we remember to get the libc dependencies back down.

**Pull request merged**
